internal/stringset/stringset.go

```diff
@@ -0,0 +1,72 @@
+// Copyright © 2018 Intel Corporation
```

Comment:
This package induces one huge WTF. WTF is this for, nothing else seems to use it. no tests...

Reply:
Great question, and you very well might come back with a great alternative implementation.

For working with bundle lists, I need a data structure to store the bundle names read in from the Mix Bundle List, to add bundles to as I recurse that list, etc.

One of the requirements for the bundle list processing is the ability to eliminate duplicate bundles.  I originally was going to just use a slice to contain the bundle names (simple, easy, built-in), but the uniqueness check then requires traversing the list each time, creating an O(n^2) operation just for insertion.

A "collection of unique items" is obviously a set. Python and Ruby both have a built-in set type, and Java has plenty of variants in their standard library. I was surprised to find Go didn't bother. The closest I found to an "official" explanation was, essentially, "If you want a set of something in Go, it's easy enough to write it yourself using a `map[something]bool` or `map[something]struct{}` for a more space-efficient variant.`

"I have to implement my own Set?!" I thought. "Surely there's a good library out there for this". And there are -- I found several -- but all contained a heck of a lot more than I needed (set intersection and union, multithread mutex locking, etc), so I gave up and wrote the 30ish (if you take out comments and whitespace) lines of code myself.

Reply:
As for why I wrote this as a generic string set, rather than a more-specific "bundle set" for a new `Bundle` data type, well -- at the time I needed it, I was only working with bundle names, which are strings.  Perhaps it'd be more useful to have a "bundle set" feature in the future. I could have called it a "bundle set" and still just made it strings, but I figured if I'm going to write a set that is simply of strings, might as well admit to that in case it's useful elsewhere for non-bundle-related strings.

As for why I made it a standalone internal package, well, that came from discussions with @cmarcelo in a few places about how best to organize generic code, and reading a bunch of Go blogs and discussions about package organization and how best to expose utility functionality.

The fact no one else uses this yet is breaking with @cmarcelo's suggestion of keeping things next to where they're used until it truly is generic. But it's in keeping with a broader opinion of "organize your utility functionality into meaningful packages with consolidated behavior".

Reply:
My point is that the 30ish lines you have written are nothing to do with this PR - "Bundle refactor". Please remove it.

Yes, go does require you to implement 'set' on top of map. map gives you the O(1) lookup time and other good properties that you typically want.

It may be that what you find up wanting in the future is a pointer to bundle, rather than a bundle.

Reply:
I feel like this knot could be untied if I've extracted the bundleset package out of the BCB patch (this is easy and wouldn't break anything), and add whatever extra things you need (you talked about also parsing some of the fields in comments). @kevincwells  Would that be helpful?

Reply:
Not sure how I feel about this addition. I see how you use it in your code and it's fine, but seems like a slightly unnecessary abstraction. I defer to @icarus-sparry and @cmarcelo  here...